apis/projectcontour/v1/httpproxy.go

```diff
@@ -171,6 +171,14 @@ type AuthorizationServer struct {
 	//
 	// +optional
 	FailOpen bool `json:"failOpen,omitempty"`
+
+	// AuthBeforeRateLimiting defines whether the authorization server
+	// will be consulted before applying any rate limiting to a
+	// request. The default is false, i.e. to apply rate-limiting first,
+	// in order to protect the authorization server.
+	//
+	// +optional
+	AuthBeforeRateLimiting bool `json:"authBeforeRateLimiting,omitempty"`
```

Comment:
I wonder if we should make this a `pritority` level or something? This would allow additional filters to be applied in different orders when/if we add them. For instance, maybe you want the lua filters to run before auth but after rate limiting. 

This priority might just be a int which the lower priority puts it higher in the list.

Reply:
There was some discussion on this in the issue (https://github.com/projectcontour/contour/issues/3726#issuecomment-987204370 and following), agree it's a more general solution but also I think coming up with a good API is pretty tricky for the reasons I mentioned in the issue which is why I didn't go down that path. If you have thoughts on how it would work across all the filters, including those configured in HTTPProxy, in the config file, and not explicitly configured at all, I'm happy to hear the ideas.

Reply:
Would you put `priority` fields in all the different places filters have configuration? E.g. in your example, global rate limiting is configured in the config file, so you'd need one priority there; auth is in `HTTPProxy.Spec.VirtualHost.Authorization`, so another there; and then Lua (assuming you're talking about the cookie rewrite stuff?) is in `HTTPProxy.Spec.Routes.CookieRewritePolicies`, so you'd need another priority there?

Reply:
> I don't think gatwayapi solves it with filters since their order isn't determined.

Gateway API does have an explicit list of filters in the API, though, so it seems easier to solve there since filters are explicit concepts that are defined together in one place, so ordering could be defined by the user either by reordering items in the list, or by using a `priority` field on the Filter structs.

Reply:
> Would you put priority fields in all the different places filters have configuration? E.g. in your example, global rate limiting is configured in the config file, so you'd need one priority there; auth is in HTTPProxy.Spec.VirtualHost.Authorization, so another there; and then Lua (assuming you're talking about the cookie rewrite stuff?) is in HTTPProxy.Spec.Routes.CookieRewritePolicies, so you'd need another priority there?

Yeah I agree this isn't great, just feels like we're immediately causing a big api issue later on by making this so specific. 

Would we need to add a `authBeforeLua` field? I guess if that was true and `authBeforeRateLimiting` was true that would be them before RateLimiting, but what order would Lua or Auth be?

Reply:
> Yeah I agree this isn't great, just feels like we're immediately causing a big api issue later on by making this so specific.

Yeah, I don't love the API implemented in this PR either, but I guess my hope was that we wouldn't need to add more filter reordering options, at least not until a v2 API where maybe we can reorganize things to make it easier.

I'll think about this some more and see if I can come up with any other options.

Reply:
something else to maybe consider: https://kubernetes.slack.com/archives/C8XRH2R4J/p1639502957273600

the newer matching API that is coming out has some interesting features (match action to skip a filter, action could be to a route, etc.)

Reply:
I think that adding in filter priority is a really big API change, big enough to warrant an API rev, and I would question whether we should invest in a HTTPProxy API rev at this point, or work on making the Gateway API support these use cases instead.

I agree with @stevesloka that having this be a boolean is limiting, but I don't think that we can solve that without a really big revamp of HTTPProxy.

Reply:
Echo this comment to build it as int if possible so we get the optionality to add more levers later on.  I can see applying certain transformations both before and after auth or rate limiting.  But leave this decision up to you guys